onnx/compose.py

```diff
@@ -0,0 +1,313 @@
+        doc_string (string): Optional docstring for the combined graph
+                             If not provided, a default docstring with the concatenation of g1 and g2 docstrings is used
+        producer_name (string): Optional producer name for the combined model. Default: 'onnx.compose'
+        producer_version (string): Optional producer version for the combined model. Default: "1.0"
+        domain (string): Optional domain of the combined model. Default: ""
+        model_version (int): Optional version of the graph encoded. Default: 1
+    """
+    if type(m1) is not ModelProto:
+        raise ValueError("m1 argument is not an ONNX model")
+    if type(m2) is not ModelProto:
+        raise ValueError("m2 argument is not an ONNX model")
+
+    if m1.ir_version != m2.ir_version:
+        raise ValueError(
+            f"IR version mismatch {m1.ir_version} != {m2.ir_version}."
+            " Both models should have have the same IR version")
+    ir_version = m1.ir_version
+
+    opset_import_map = {}  # type: MutableMapping[Text, int]
+    opset_imports = \
+        [entry for entry in m1.opset_import] + \
+        [entry for entry in m2.opset_import]
+
+    for entry in opset_imports:
+        if entry.domain in opset_import_map:
+            found_version = opset_import_map[entry.domain]
+            if entry.version != found_version:
+                raise ValueError(
+                    "Can't merge two models with different operator set ids for a given domain. "
+                    f"Got: {m1.opset_import} and {m2.opset_import}")
+        else:
+            opset_import_map[entry.domain] = entry.version
+
+    graph = merge_graphs(m1.graph, m2.graph, io_map, name, doc_string)
+    model = helper.make_model(graph,
+                              producer_name=producer_name,
+                              producer_version=producer_version,
+                              domain=domain,
+                              model_version=model_version,
+                              opset_imports=opset_imports,
+                              ir_version=ir_version)
+
+    # Merging model metadata props
+    model_props = {}
+    for meta_entry in m1.metadata_props:
+        model_props[meta_entry.key] = meta_entry.value
+    for meta_entry in m2.metadata_props:
+        if meta_entry.key in model_props:
+            value = model_props[meta_entry.key]
+            if value != meta_entry.value:
+                raise ValueError(
+                    "Can't merge models with different values for the same model metadata property."
+                    f" Found: property = {meta_entry.key}, with values {value} and {meta_entry.value}."
+                )
+        else:
+            model_props[meta_entry.key] = meta_entry.value
+    helper.set_model_props(model, model_props)
+
+    # Merging functions
+    model.functions.MergeFrom(m1.functions)
+    model.functions.MergeFrom(m2.functions)
```

Comment:
There could be functions in the two models that use the same name, which would lead to a conflict here.

Reply:
I've added a name collision check. Is there anything else that can be done? Should we rename local functions, or just leave it as a limitation?

Reply:
I've now added renaming of local functions, including tests

Reply:
Great, thanks very much!